### PR TITLE
[wip] Add PermissionsTest

### PR DESCRIPTION
**Pull request by samsonasik**
@MGatner please verify if this is the way to apply permission to user id, thank you.

### CODE REVIEW

**Comment by MGatner** [comment]
Not quite. `PermitsTrait` works on the target model. I will make the first test or two so you get the idea - standby.

**Comment by samsonasik** [comment]
ok, thank you ;)

**Comment by MGatner** [comment]
@samsonasik: https://github.com/tattersoftware/codeigniter4-files/pull/14

**Comment by samsonasik** [comment]
Thank you

**Comment by samsonasik** [comment]
I cherry-picked your commit

**Comment by MGatner** [comment]
Sounds good. Be sure to grab 203fe6bd00d4acdabc833a461232cc6d7b85a163 as well - I messed up the permission names originally. That also has an example of logging a user in to check authenticated permissions.

**Comment by samsonasik** [comment]
Ok, I've cherry-picked https://github.com/tattersoftware/codeigniter4-files/commit/203fe6bd00d4acdabc833a461232cc6d7b85a163

**Comment by samsonasik** [comment]
phpstan seems errored at https://github.com/tattersoftware/codeigniter4-files/pull/14 :

```bash

Run vendor/bin/phpstan analyze
Note: Using configuration file /home/runner/work/codeigniter4-files/codeigniter4-files/phpstan.neon.dist.
Killed
Error: Process completed with exit code 137.
```

It seems we can disable phpstan check on tests.

File: tests/_support/FilesTestCase.php [hunk 1]
```diff
@@ -103,7 +103,7 @@ protected function tearDown(): void
 	protected function login(int $userId = null): User
 	{
 		// Get or create the user
-		$user = $userId ? model(UserModel::class)->find($userId) : fake(UserFaker::class);
+		$user = $userId ? model(Models\UserModel::class)->find($userId) : fake(UserFaker::class);
```
Comment (samsonasik) [inline tests/_support/FilesTestCase.php:106]: catched by phpstan, fixed now.

**Comment by samsonasik** [comment]
@MGatner I added `testAuthenticatedListOwnOnly` and `testAdminAccess`. 

I am not sure about "Authenticated adds only", it seems for controller action "upload" and "bulk" ? Is there need `mayCreate()` check like this:

```php
	public function upload()
	{
		// Check for create permission
		if (! $this->model->mayCreate())
		{
			return $this->user();
		}
                // ...
         }
```

I tried add above `mayCreate()`, but testing seems not working for test "empty" upload which should get 400 status:

```php
	public function testAuthenticatedAddOnly()
	{
		$this->setMode(00222);
		$this->login($this->user->id);

		$result = $this->withSession()->get('files/upload');
		$result->assertStatus(400);
	}
```

Tried with user got 200 status, with admin got 302 status.

**Comment by MGatner** [comment]
I think you've done it correct but set the wrong mode. `2` is the write bit, so `00222` means "nobody list, everybody writes". Try the same code with `setMode(06660)`

**Comment by MGatner** [comment]
Or probably even better would be `04664` since I imagine that's what most implementations will want to use.

**Comment by samsonasik** [comment]
I just added `testAuthenticatedAddOnly` with add `mayCreate()` check in upload action, it got error: 

````bash
➜  codeigniter4-files git:(permission-test) vendor/bin/phpunit tests/feature/PermissionsTest.php --filter testAuthenticatedAddOnly
PHPUnit 8.5.8 by Sebastian Bergmann and contributors.

Error:         No code coverage driver is available

F                                                                   1 / 1 (100%)

Time: 1.88 seconds, Memory: 14.00 MB

There was 1 failure:

1) PermissionsTest::testAuthenticatedAddOnly
Failed asserting that 302 matches expected 400.

/Users/samsonasik/www/codeigniter4-files/vendor/codeigniter4/codeigniter4/system/Test/FeatureResponse.php:169
/Users/samsonasik/www/codeigniter4-files/tests/feature/PermissionsTest.php:115
```

**Comment by MGatner** [comment]
Also since `upload` will usually be an AJAX endpoint I think you will need to check for it and use `failure` instead of `user` when the permissions fail.

**Comment by samsonasik** [comment]
Thank you, it seems fixed now, use no file got expected 400 status.

**Comment by samsonasik** [comment]
@MGatner do you have example how to tests file upload that make `upload->isValid() === true` ?

**Comment by MGatner** [comment]
I haven't done it before. You might check out **CodeIgniter4/tests/system/HTTP/Files/FileCollectionTest.php**. I'm also happy to find a way to inject test results into the controller, like `setData(['isValid' => true])` or something.

**Comment by samsonasik** [comment]
Ok ,I will try

**Comment by samsonasik** [comment]
implemented upload file succeed test with Patchwork http://patchwork2.org/ with redefine function on the fly.

File: tests/feature/PermissionsTest.php [hunk 2]
```diff
@@ -0,0 +1,242 @@
...
+
+		$fileOwnByAdmin = fake(FileFaker::class);
+		model('FileModel')->addToUser($fileOwnByAdmin->id, $this->admin->id);
+
+		$result = $this->withSession()->get('files/user/' . $this->proctor->id);
+		$result->assertStatus(200);
+
+		$files = $this->model->getForUser($this->proctor->id);
+		$result->assertSee($fileOwnByProctor->filename);
+		$result->assertSee($fileOwnByProctor2->filename);
```
Comment (MGatner) [inline tests/feature/PermissionsTest.php:216]: A user visiting `files/user/{their_own_id}` should only see their own files. Something is wrong if another user's file is showing up.

Comment (samsonasik) [reply tests/feature/PermissionsTest.php:216]: Yes, both 2 line are for addition to proctor user, in next line, there is :
 ```php
$result->assertDontSee($fileOwnByAdmin->filename)
````
 for **admin** that proctor can't see admin file, is that already correct ?

Comment (MGatner) [reply tests/feature/PermissionsTest.php:216]: Ohhhh yes you are right, I misread that as "file owned by proctor 2 {a different user}". Looks good!

**Review by MGatner — APPROVED** [review]
